### DISCUSSION TITLE
[@types/amplitude-js] is missing cookieStorage

### DISCUSSION THREAD

**Discussion by bernardoduarte** [DefinitelyTyped/DefinitelyTyped#57387] [Issues with a @types package]
The amplitude client at https://github.com/DefinitelyTyped/DefinitelyTyped/blob/5bdfa971b6a8422f8fe1d7ea2a87f1f1e911a8fe/types/amplitude-js/index.d.ts#L90 is missing a valid attribute `cookieStorage`, that is available via the javascript SDK using their html script tag.

A workaround that worked for me is to add a interface that created a map for this attribute:

```typescript
interface LegalcloudAmplitudeClient extends AmplitudeClient {
    cookieStorage: Map<any, any>;      
}
```

[a comment by an automated account is not shown]

**Comment by danmana** [comment]
Hi @bernardoduarte

Is the `cookieStorage` part of the public API of AplitudeClient?

I don't see it mentioned anywhere in the docs: https://amplitude.github.io/Amplitude-JavaScript/
Though, by the naming convention, it might be considered a public property (no leading underscore): https://github.com/amplitude/Amplitude-JavaScript/blob/v8.13.0/src/amplitude-client.js#L41

What is the use-case for it? AmplitudeClient seems to do all the handling it needs internally.

At a first glance, it looks like it would not be complicated to add some typing for it.
However, I'm not a user of Amplitude at the moment, so I don't know if working with the `cookieStorage` is something that is usually recommended or not. 

@asido @HintikkaKimmo what do you think?

**Comment by bernardoduarte** [comment]
Hello there @danmana ;D

> Is the cookieStorage part of the public API of AplitudeClient?

AFAIK it is, because it can be accessed through the SDK:

```
<script type="text/javascript">
    (function(e,t){var n=e.amplitude||{_q:[],_iq:{}};var r=t.createElement("script")
    ;r.type="text/javascript"
    ;r.integrity="sha384-tzcaaCH5+KXD4sGaDozev6oElQhsVfbJvdi3//c2YvbY02LrNlbpGdt3Wq4rWonS"
    ;r.crossOrigin="anonymous";r.async=true
    ;r.src="https://cdn.amplitude.com/libs/amplitude-8.5.0-min.gz.js"
    ;r.onload=function(){if(!e.amplitude.runQueuedFunctions){
    console.log("[Amplitude] Error: could not load SDK")}}
    ;var i=t.getElementsByTagName("script")[0];i.parentNode.insertBefore(r,i)
    ;function s(e,t){e.prototype[t]=function(){
    this._q.push([t].concat(Array.prototype.slice.call(arguments,0)));return this}}
    var o=function(){this._q=[];return this}
    ;var a=["add","append","clearAll","prepend","set","setOnce","unset","preInsert","postInsert","remove"]
    ;for(var c=0;c<a.length;c++){s(o,a[c])}n.Identify=o;var u=function(){this._q=[]
    ;return this}
    ;var l=["setProductId","setQuantity","setPrice","setRevenueType","setEventProperties"]
    ;for(var p=0;p<l.length;p++){s(u,l[p])}n.Revenue=u
    ;var d=["init","logEvent","logRevenue","setUserId","setUserProperties","setOptOut","setVersionName","setDomain","setDeviceId","enableTracking","setGlobalUserProperties","identify","clearUserProperties","setGroup","logRevenueV2","regenerateDeviceId","groupIdentify","onInit","logEventWithTimestamp","logEventWithGroups","setSessionId","resetSessionId"]
    ;function v(e){function t(t){e[t]=function(){
    e._q.push([t].concat(Array.prototype.slice.call(arguments,0)))}}
    for(var n=0;n<d.length;n++){t(d[n])}}v(n);n.getInstance=function(e){
    e=(!e||e.length===0?"$default_instance":e).toLowerCase()
    ;if(!Object.prototype.hasOwnProperty.call(n._iq,e)){n._iq[e]={_q:[]};v(n._iq[e])
    }return n._iq[e]};e.amplitude=n})(window,document);
    
    amplitude.getInstance().init("YOUR_KEY");
    var deviceId = amplitude.getInstance().cookieStorage.get('deviceId');
    if (deviceId) {
        amplitude.getInstance().setDeviceId(deviceId);
    } else {
        amplitude.getInstance().cookieStorage.set('deviceId', amplitude.getInstance().options.deviceId);
    }
</script>
```

I'm using it to enable [cross-domain tracking](https://help.amplitude.com/hc/en-us/articles/115003135607-Track-unique-users-in-Amplitude#h_f717e48c-460c-4a17-abd7-102327673798), and it works perfectly when using the aforementioned SDK.

**Comment by danmana** [comment]
Added a PR https://github.com/DefinitelyTyped/DefinitelyTyped/pull/57578
